src/crafting_gui.cpp

```diff
@@ -1338,14 +1395,76 @@ const recipe *select_crafting_recipe( int &batch_size_out, const recipe_id goto_
         just_toggled_unread = false;
         ui_manager::redraw();
         const int scroll_item_info_lines = catacurses::getmaxy( w_iteminfo ) - 4;
-        const std::string action = ctxt.handle_input();
+        std::string action = ctxt.handle_input();
         const int recmax = static_cast<int>( current.size() );
         const int scroll_rate = recmax > 20 ? 10 : 3;
-        if( action == "SCROLL_RECIPE_INFO_UP" ) {
+
+        cata::optional<tripoint> coord3d = ctxt.get_coordinates( catacurses::stdscr );
+        cata::optional<point> coord;
+        if( coord3d.has_value() ) {
+            // TODO: Handle this in get_coordinates, which currently only normalizes for w_terrain
+            coord = point( coord3d->x + TERMX / 2, coord3d->y + TERMY / 2 );
+        }
+        const bool mouse_in_list = coord.has_value() && mouseover_area_list.contains( coord.value() );
+        const bool mouse_in_recipe = coord.has_value() && mouseover_area_recipe.contains( coord.value() );
+
+        // Check mouse selection of recipes separately so that selecting an already-selected recipe
+        // can go straight to "CONFIRM"
+        if( action == "SELECT" ) {
+            if( !list_map.empty() && coord.has_value() ) {
+                point local_coord = coord.value() - point( getbegx( w_data ), getbegy( w_data ) );
+                for( const auto &entry : list_map ) {
+                    if( entry.second.contains( local_coord ) ) {
+                        if( line == static_cast<int>( entry.first ) ) {
+                            action = "CONFIRM";
+                        } else {
+                            if( !previously_toggled_unread ) {
+                                last_line = line;
+                            }
+                            line = entry.first;
+                            user_moved_line = highlight_unread_recipes;
+                        }
+                    }
+                }
+            }
+        }
+
+        if( action == "SELECT" ) {
+            bool handled = false;
+            if( coord.has_value() ) {
+                point local_coord = coord.value() - point( getbegx( w_head_tabs ), getbegy( w_head_tabs ) );
+                if( !translated_tab_map.empty() ) {
```

Comment:
You don't have to check this before iterating over the a container. The loop works fine even if the container is empty.